PolyEngine/Core/Src/BTreeMap.hpp

```diff
@@ -0,0 +1,532 @@
+		Optional<V> Insert(      K&& key, const V&  value) { return this->insert(std::move(key),           value ); }
+		Optional<V> Insert(      K&& key,       V&& value) { return this->insert(std::move(key), std::move(value)); }
+
+		Optional<V> Remove(const K&  key) { return this->remove(key); }
+		Optional<V> Remove(      K&& key) { return this->remove(std::move(key)); }
+
+		Optional<V&> Get(const K&  key) { return this->get(key); }
+		Optional<V&> Get(      K&  key) { return this->get(std::move(key)); }
+
+		V& operator[](const K& key) { return this->Get(key).value(); }
+		size_t GetSize() { return this->len; };
+		bool IsEmpty()   { return this->GetSize() == 0; };
+
+		void Clear() {
+			KVERef current = BTree::first_leaf_edge(this->root.as_node_ref());
+			while (this->len) {
+				this->len -= 1;
+				if (current.idx < current.node_ref.node->len) {
+					current = KVERef{current.node_ref, current.idx + 1};
+					continue;
+				} else {
+					const NodeRef to_delete = current.node_ref;
+					current = current.node_ref.ascend().parent;
+					delete to_delete.node;
+				}
+
+				for (;;) {
+					if (current.idx < current.node_ref.node->len) {
+						current = BTree::first_leaf_edge(KVERef{current.node_ref, current.idx + 1}.descend());
+						break; //continue outer loop
+					} else {
+						const NodeRef to_delete = current.node_ref;
+						current = current.node_ref.ascend().parent;
+						delete to_delete.node->as_branch();
+					}
+				}
+			}
+
+			LeafNode* new_root_to_be = current.node_ref.node;
+
+			auto ascension = current.node_ref.ascend();
+			while (ascension.succeeded) {
+				KVERef node = ascension.parent;
+				ascension = node.node_ref.ascend();
+				delete node.node_ref.node->as_branch();
+			}
+
+			//reinitialize the root node
+			new_root_to_be->~LeafNode();
+			::new(new_root_to_be) LeafNode();
+
+			this->root = {new_root_to_be, 0};
+		}
+
+	public:
+		Entry Entry(const K& key) {
+			const auto search_result = search_tree(this->root.as_node_ref(), key);
+			switch (search_result.result) {
+				case SearchResult::FOUND:  return Entry::Occupied(search_result.handle, this->len);
+				case SearchResult::FAILED: return Entry::Vacant(  search_result.handle, this->len, key);
+				default: UNREACHABLE();
+			}
+		}
+		Range<K, V, Bfactor> MaxRange() { return {BTree::first_leaf_edge(this->root.as_node_ref()), BTree::last_leaf_edge(this->root.as_node_ref())}; }
+		Range<K, V, Bfactor> GetRange(const K& lower, const K& upper) { ASSERTE(lower < upper, ""); return {search_tree(this->root.as_node_ref(), lower).handle, search_tree(this->root.as_node_ref(), upper).handle}; }
+		Iter begin() { return {BTree::first_leaf_edge(this->root.as_node_ref()), this->GetSize(), this->GetSize()}; }
+		Iter end()   { return {BTree:: last_leaf_edge(this->root.as_node_ref()), 0,               this->GetSize()}; }
+		class Keys   Keys()   { return {*this}; }
+		class Values Values() { return {*this}; }
```

Comment:
Provide const overloads for these six methods. `Keys()` and `Values()` should return propper iterator ranges. Ofcourse that implies implementation of `KeyIterator` and `ValueIterator`.